### DISCUSSION TITLE
How to list packages and versions in github package registry?

### DISCUSSION THREAD

**Discussion by tombrus** [community/community#24680] [Apps, API and Webhooks]
<p>What is the best way to list github packages in a script?</p>
<p>Thanks,</p>
<p>Tom</p>

**Comment by tombrus** [comment]
<p>…after a little bit of hacking I came up with the following bash script, using GraphQL/curl/jq:</p>
<pre><code>graphqlQuery() {
  local query="$1"; shift

  curl -s -H "Authorization: bearer $INPUT_TOKEN" -X POST -d '{"query":"'"$query"'"}' 'https://api.github.com/graphql'
}
listPackageVersions() {
  local g="$1"; shift
  local a="$1"; shift

  local query="$(cat &lt;&lt;EOF | sed 's/"/\\"/g' | tr '\n\r' ' '
query {
    repository(owner:"$USERNAME", name:"$REPOSNAME"){
        registryPackages(name:"$g.$a",first:1) {
            nodes {
                versions(last:100) {
                    nodes {
                        version
                    }
                }
            }
        }
    }
}
EOF
)"
  graphqlQuery "$query" | jq -r '.data.repository.registryPackages.nodes[0].versions.nodes[].version'
}
</code></pre>
<p>just thought I’d share it.</p>
<p>-Tom</p>

**Comment by tombrus** [comment]
<p>Since the <code>registryPackages</code> seems to have been refactored to <code>packages</code> I had to change my script around. Again, just thought I’d share it… her is my new version:</p>
<pre><code class="lang-auto">graphqlQuery() {
  local query="$1"; shift

  curl -s -H "Authorization: bearer $INPUT_TOKEN" -X POST -d '{"query":"'"$query"'"}' 'https://api.github.com/graphql'
}
listPackageVersions() {
  local g="$1"; shift
  local a="$1"; shift

  local query="$(cat &lt;&lt;EOF | sed 's/"/\\"/g' | tr '\n\r' ' '
query {
    repository(owner:"$USERNAME", name:"$REPOSNAME"){
        packages(names:"$g.$a",first:1) {
            nodes {
                versions(first:100) {
                    nodes {
                        version
                    }
                }
            }
        }
    }
}
EOF
)"
  graphqlQuery "$query" | jq -r '.data.repository.packages.nodes[0].versions.nodes[].version'
}
</code></pre>

**Comment by Fronix** [comment]
<p>Apparantly they removed the updatedAt aswell which is what I was using… Not sure how to get that date now can’t find anything.</p>
<p>Following does not work:</p>
<pre><code class="lang-auto">query {
          repository(owner: "User", name: "repo") {
              packages(first:100){
                nodes {
                    name,
                    packageType 
                    versions(first:100) {
                        nodes {
                            id,
                            version,
                            updatedAt
                        }
                    }
                }
              }
          }
      }
</code></pre>

**Comment by Fronix** [comment]
<p>So it seems the only way to get a date that should be something close to the actual package date would be <code>PackageFile</code> object.</p>
<p>New query:</p>
<pre><code class="lang-auto">query {
          repository(owner: "User", name: "repo") {
              packages(first:100){
                nodes {
                    id
                    name
                    packageType 
                    versions(first:100) {
                        nodes {
                            id
                            version
                            files(first: 10) {
                                nodes {
                                    name
                                    updatedAt
                                }
                            }
                        }
                    }
                }
              }
          }
      }
</code></pre>

**Comment by tombrus** [comment, marked as answer]
<aside class="quote no-group" data-username="fronix" data-post="4" data-topic="16999">
<div class="title">
<div class="quote-controls"></div>
 fronix:</div>
<blockquote>
<p>updatedAt</p>
</blockquote>
</aside>
<p>I found the following on  the <a href="https://developer.github.com/v4/" rel="nofollow noopener">GraphQL API v4</a> change log page:</p>
<p>reg your <em>updatedAt</em>:</p>
<pre><code class="lang-nohighlight">On member RegistryPackageVersion.updatedAt: updatedAt will be removed. 
Use the PackageVersion object instead. Effective 2020-04-01.
</code></pre>
<p>and on my <em>registryPackages</em>:</p>
<pre><code class="lang-nohighlight">On member Organization.registryPackages: registryPackages will be removed.
Use the PackageOwner object instead. Effective 2020-04-01.
</code></pre>
<p>So I guess we are using some moving targets :frowning:<br>
On the other hand, improving APIs is also good! :slight_smile:</p>

**Comment by Fronix** [comment]
<p>Yeah I found that as well, well seems like we’ll have to rewrite these things a few times before its out of beta :stuck_out_tongue:</p>

**Comment by gihanrangana** [comment]
<p>how can i run this on nodejs? i have no idea about GraphQl</p>

**Comment by joshjohanning** [comment]
<p><a class="mention" href="/sl-codeblaster">@SL-CodeBlaster</a> - (late reply but for others who find this)</p>
<p>You can use the GitHub GraphQL Explorer: <a href="https://docs.github.com/en/graphql/overview/explorer" class="inline-onebox" rel="noopener nofollow ugc">Explorer - GitHub Docs</a></p>
<p>Or you could use Postman:</p>
<ol>
<li>write a <code>POST</code> to <code>https://api.github.com/graphql</code>
</li>
<li>Add your GitHub PAT under the Authorization tab as a <code>Bearer</code>
</li>
<li>In the Body, change to <code>GraphQL</code> and copy in the query here</li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/0/2/02e77457d7880cd59248cd9a751c263e421a67a1.png" data-download-href="/uploads/short-url/pGQjK6Q91VJugybQdb7D8ekwQ9.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/19912012/181086036-71372bca-96b5-493e-a8aa-4b049e636227.png" alt="image" data-base62-sha1="pGQjK6Q91VJugybQdb7D8ekwQ9" width="690" height="269" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/0/2/02e77457d7880cd59248cd9a751c263e421a67a1_2_690x269.png, https://aws1.discourse-cdn.com/github/original/3X/0/2/02e77457d7880cd59248cd9a751c263e421a67a1.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/0/2/02e77457d7880cd59248cd9a751c263e421a67a1.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/0/2/02e77457d7880cd59248cd9a751c263e421a67a1_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">938×366 39.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/d/d/ddd6f412eed597d78fe9e2273a20100af6627c65.png" data-download-href="/uploads/short-url/vEu3vjLn9rBiUewzccBLDPnKvNH.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/19912012/181086037-5aa9855e-538d-411d-93b0-57cfbcc7c81b.png" alt="image" data-base62-sha1="vEu3vjLn9rBiUewzccBLDPnKvNH" width="690" height="425" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/d/ddd6f412eed597d78fe9e2273a20100af6627c65_2_690x425.png, https://aws1.discourse-cdn.com/github/original/3X/d/d/ddd6f412eed597d78fe9e2273a20100af6627c65.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/d/d/ddd6f412eed597d78fe9e2273a20100af6627c65.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/d/ddd6f412eed597d78fe9e2273a20100af6627c65_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">978×603 42.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by ubudragon** [comment]
doesn't work anymore, pakageType npm deprecated and we must use GitHub rest API now

**Reply by joshjohanning** [reply]
```bash
packages=$(gh api --paginate "/orgs/${ORG}/packages?package_type=npm" -q '.[] | .name + " " + .repository.name')

echo "$packages" | while IFS= read -r response; do

  package_name=$(echo "$response" | cut -d ' ' -f 1)
  repo_name=$(echo "$response" | cut -d ' ' -f 2)

  echo "org: $ORG repo: $repo_name --> package name $package_name"
  
  versions=$(gh api --paginate "/orgs/${ORG}/packages/npm/${package_name}/versions" -q '.[] | .name' | sort -V)

  for version in $versions
  do
    echo "$version"
  done
```

https://docs.github.com/en/free-pro-team@latest/rest/packages/packages?apiVersion=2022-11-28#list-packages-for-an-organization

**Reply by ubudragon** [reply]
thanks, joshjohanning!
but for our org, it returns only 30 packages instead of 150+ total
I tried to call the gh API with a different page: headers - all the same, only 30 packages in response (despite the response json has nearly 3000 lines)

**Reply by joshjohanning** [reply]
ahh oops forgot pagination @ubudragon! Edited the above code snippet. 

Need the `gh api --paginate` to retrieve multiple pages of results 😄
